File: tests/qt_types_standalone/rust/src/qvariant.rs
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
// SPDX-FileContributor: Gerhard de Clercq <gerhard.declercq@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0
use cxx_qt_lib::{
    QByteArray, QColor, QDate, QDateTime, QPoint, QPointF, QRect, QRectF, QSize, QSizeF, QString,
    QTime, QTimeZone, QUrl, QVariant,
};

#[cxx::bridge]
mod qvariant_cxx {
    enum VariantTest {
        Bool,
        F32,
        F64,
        I8,
        I16,
        I32,
        QByteArray,
        QColor,
        QDate,
        QDateTime,
        QPoint,
        QPointF,
        QRect,
        QRectF,
        QSize,
        QSizeF,
        QTime,
        QUrl,
        QString,
        U8,
        U16,
        U32,
    }

    unsafe extern "C++" {
        include!("cxx-qt-lib/qvariant.h");

        type QVariant = cxx_qt_lib::QVariant;
    }

    extern "Rust" {
        fn construct_qvariant(test: VariantTest) -> QVariant;
        fn read_qvariant(v: &QVariant, test: VariantTest) -> bool;
        fn clone_qvariant(v: &QVariant) -> QVariant;
    }
}

use qvariant_cxx::VariantTest;

fn construct_qvariant(test: VariantTest) -> QVariant {
    match test {
        VariantTest::Bool => QVariant::from(&true),
        VariantTest::F32 => QVariant::from(&1.23_f32),
        VariantTest::F64 => QVariant::from(&1.23_f64),
        VariantTest::I8 => QVariant::from(&12_i8),
        VariantTest::I16 => QVariant::from(&123_i16),
        VariantTest::I32 => QVariant::from(&123_i32),
        VariantTest::QByteArray => QVariant::from(&QByteArray::from("Rust bytes")),
        VariantTest::QColor => QVariant::from(&QColor::from_rgb(255, 0, 0)),
        VariantTest::QDate => QVariant::from(&QDate::new(2022, 1, 1)),
        VariantTest::QDateTime => QVariant::from(&QDateTime::from_date_and_time_time_zone(
            &QDate::new(2022, 1, 1),
            &QTime::new(1, 2, 3, 4),
            &QTimeZone::owned_from_offset_seconds(0),
        )),
        VariantTest::QPoint => QVariant::from(&QPoint::new(1, 3)),
        VariantTest::QPointF => QVariant::from(&QPointF::new(1.0, 3.0)),
        VariantTest::QRect => QVariant::from(&QRect::new(123, 456, 246, 912)),
        VariantTest::QRectF => QVariant::from(&QRectF::new(1.23, 4.56, 2.46, 9.12)),
        VariantTest::QSize => QVariant::from(&QSize::new(1, 3)),
        VariantTest::QSizeF => QVariant::from(&QSizeF::new(1.0, 3.0)),
        VariantTest::QString => QVariant::from(&QString::from("Rust string")),
        VariantTest::QTime => QVariant::from(&QTime::new(1, 2, 3, 4)),
        VariantTest::QUrl => QVariant::from(&QUrl::from("https://github.com/KDAB")),
        VariantTest::U8 => QVariant::from(&12_u8),
        VariantTest::U16 => QVariant::from(&123_u16),
        VariantTest::U32 => QVariant::from(&123_u32),
        _others => panic!("Unsupported test: {}", test.repr),
    }
}

fn read_qvariant(v: &cxx_qt_lib::QVariant, test: VariantTest) -> bool {
    match test {
        VariantTest::Bool => match v.value::<bool>() {
            Some(b) => !b,
            None => false,
        },
        VariantTest::F32 => match v.value::<f32>() {
            Some(f) => f == 89.1,
            None => false,
        },
        VariantTest::F64 => match v.value::<f64>() {
            Some(f) => f == 89.1,
            None => false,
        },
        VariantTest::I8 => match v.value::<i8>() {
            Some(i) => i == 89,
            None => false,
        },
        VariantTest::I16 => match v.value::<i16>() {
            Some(i) => i == 8910,
            None => false,
        },
        VariantTest::I32 => match v.value::<i32>() {
            Some(i) => i == 8910,
            None => false,
        },
        VariantTest::QByteArray => match v.value::<QByteArray>() {
            Some(bytes) => bytes.to_string() == "C++ bytes",
            None => false,
        },
        VariantTest::QColor => match v.value::<QColor>() {
            Some(color) => {
                color.alpha() == 255
                    && color.red() == 0
                    && color.green() == 255
                    && color.blue() == 0
            }
            None => false,
        },
        VariantTest::QDate => match v.value::<QDate>() {
            Some(date) => date.year() == 2021 && date.month() == 12 && date.day() == 31,
            None => false,
        },
        VariantTest::QDateTime => match v.value::<QDateTime>() {
            Some(date_time) => {
                date_time.date().year() == 2021
                    && date_time.date().month() == 12
                    && date_time.date().day() == 31
                    && date_time.time().hour() == 4
                    && date_time.time().minute() == 3
                    && date_time.time().second() == 2
                    && date_time.time().msec() == 1
                    && date_time.offset_from_utc() == 0
            }
            None => false,
        },
        VariantTest::QPoint => match v.value::<QPoint>() {
            Some(point) => point.x() == 8 && point.y() == 9,
            None => false,
        },
        VariantTest::QPointF => match v.value::<QPointF>() {
            Some(pointf) => pointf.x() == 8.0 && pointf.y() == 9.0,
            None => false,
        },
        VariantTest::QRect => match v.value::<QRect>() {
            Some(rect) => {
                rect.x() == 123 && rect.y() == 456 && rect.width() == 246 && rect.height() == 912
            }
            None => false,
        },
        VariantTest::QRectF => match v.value::<QRectF>() {
            Some(rectf) => {
                ((rectf.x() - 1.23).abs() < f64::EPSILON)
                    && ((rectf.y() - 4.56).abs() < f64::EPSILON)
                    && ((rectf.width() - 2.46).abs() < f64::EPSILON)
                    && ((rectf.height() - 9.12).abs() < f64::EPSILON)
            }
            None => false,
        },
        VariantTest::QSize => match v.value::<QSize>() {
            Some(size) => size.width() == 8 && size.height() == 9,
            None => false,
        },
        VariantTest::QSizeF => match v.value::<QSizeF>() {
            Some(sizef) => sizef.width() == 8.0 && sizef.height() == 9.0,
            None => false,
        },
        VariantTest::QString => match v.value::<QString>() {
            Some(s) => s.to_string() == "C++ string",
            None => false,
        },
        VariantTest::QTime => match v.value::<QTime>() {
            Some(time) => {
                time.hour() == 4 && time.minute() == 3 && time.second() == 2 && time.msec() == 1
            }
            None => false,
        },
        VariantTest::QUrl => match v.value::<QUrl>() {
            Some(url) => url.to_string() == "https://github.com/KDAB/cxx-qt",
            None => false,
        },
        VariantTest::U8 => match v.value::<u8>() {
            Some(i) => i == 89,
            None => false,
        },
        VariantTest::U16 => match v.value::<u16>() {
            Some(i) => i == 8910,
            None => false,
        },
        VariantTest::U32 => match v.value::<u32>() {
            Some(i) => i == 8910,
            None => false,
        },
        _others => panic!("Unsupported test: {}", test.repr),
    }
}

fn clone_qvariant(v: &QVariant) -> QVariant {
    v.clone()
}

File: crates/cxx-qt-lib/src/types/qvariant.cpp
// clang-format off
// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
// clang-format on
// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
// SPDX-FileContributor: Leon Matthes <leon.matthes@kdab.com>
//
// SPDX-License-Identifier: MIT OR Apache-2.0
#include "cxx-qt-lib/qvariant.h"

#include <QMetaObject>

#include "assertion_utils.h"

// The layout has changed between Qt 5 and Qt 6
//
// Qt5 QVariant has one member, which contains three uints and a union.
// The three uints are optimised to a reduced size, resulting in a combined size
// of two pointers.
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/kernel/qvariant.h?h=v5.15.6-lts-lgpl#n491
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/kernel/qvariant.h?h=v5.15.6-lts-lgpl#n411
//
// Qt6 QVariant has one member, which contains three pointers and a union
// (with a pointer as the largest member)
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/kernel/qvariant.h?h=v6.2.4#n540
// https://code.qt.io/cgit/qt/qtbase.git/tree/src/corelib/kernel/qvariant.h?h=v6.2.4#n474
#if (QT_VERSION >= QT_VERSION_CHECK(6, 0, 0))
assert_alignment_and_size(QVariant,
                          alignof(std::size_t),
                          sizeof(std::size_t[4]));
#else
assert_alignment_and_size(QVariant,
                          alignof(std::size_t),
                          sizeof(std::size_t[2]));
#endif

static_assert(!std::is_trivially_copy_assignable<QVariant>::value);
static_assert(!std::is_trivially_copy_constructible<QVariant>::value);

// Ensure that trivially destructible is correct
// If this is false then we need to manually implement Drop rather than derive
static_assert(!std::is_trivially_destructible<QVariant>::value);

static_assert(QTypeInfo<QVariant>::isRelocatable);

namespace rust {
namespace cxxqtlib1 {

types::QVariantType
qvariantType(const QVariant& variant)
{
#if (QT_VERSION >= QT_VERSION_CHECK(6, 0, 0))
  switch (static_cast<QMetaType::Type>(variant.metaType().id())) {
#else
  // QVariant::Type is obsolete, ensure we use QMetaType::Type to avoid
  // warnings
  switch (static_cast<QMetaType::Type>(variant.type())) {
#endif
    case QMetaType::Bool:
      return types::QVariantType::Bool;
    case QMetaType::Float:
      return types::QVariantType::F32;
    case QMetaType::Double:
      return types::QVariantType::F64;
    case QMetaType::SChar:
      return types::QVariantType::I8;
    case QMetaType::Short:
      return types::QVariantType::I16;
    case QMetaType::Int:
      return types::QVariantType::I32;
    case QMetaType::QColor:
      return types::QVariantType::QColor;
    case QMetaType::QDate:
      return types::QVariantType::QDate;
    case QMetaType::QDateTime:
      return types::QVariantType::QDateTime;
    case QMetaType::QPoint:
      return types::QVariantType::QPoint;
    case QMetaType::QPointF:
      return types::QVariantType::QPointF;
    case QMetaType::QRect:
      return types::QVariantType::QRect;
    case QMetaType::QRectF:
      return types::QVariantType::QRectF;
    case QMetaType::QSize:
      return types::QVariantType::QSize;
    case QMetaType::QSizeF:
      return types::QVariantType::QSizeF;
    case QMetaType::QString:
      return types::QVariantType::QString;
    case QMetaType::QTime:
      return types::QVariantType::QTime;
    case QMetaType::QUrl:
      return types::QVariantType::QUrl;
    case QMetaType::UChar:
      return types::QVariantType::U8;
    case QMetaType::UShort:
      return types::QVariantType::U16;
    case QMetaType::UInt:
      return types::QVariantType::U32;

    default:
      return types::QVariantType::Unsupported;
  }
}

} // namespace cxxqtlib1
} // namespace rust
